### ISSUE TITLE
Need help on make

### ISSUE THREAD

**Issue by officialmiket** [VolatileMark/opentrack-wine-guide#1]
I have done everything on the cmake but when i type make -j$(nproc) it gives me 

[ 48%] Generating opentrack-wrapper-wine.exe.so
[ 52%] Built target opentrack-options
[ 52%] Built target i18n-lrelease
In file included from /home/burger/Documents/opentrack-opentrack-2021.2.0/proto-wine/opentrack-wrapper-wine-main.cxx:15:
/home/burger/Documents/opentrack-opentrack-2021.2.0/compat/shm.h:10:10: fatal error: windows.h: No such file or directory
   10 | #include <windows.h>
      |          ^~~~~~~~~~~
compilation terminated.
winegcc: /usr/bin/g++ failed
make[2]: *** [proto-wine/CMakeFiles/wine-wrapper.dir/build.make:82: proto-wine/opentrack-wrapper-wine.exe.so] Error 2
make[1]: *** [CMakeFiles/Makefile2:2452: proto-wine/CMakeFiles/wine-wrapper.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 53%] Built target opentrack-csv
[ 54%] Built target opentrack-video
[ 56%] Built target opentrack-spline
make: *** [Makefile:149: all] Error 2

If someone would like to help me and tell me how to fix it i would appreciative. Should i post this also on opentrack github?

**Comment by SkrapeProjects** [comment]
Mmmmm.... Seems like you're missing the something from the wine development package. What distro are you using? Could you also check if `/usr/include/wine/windows/windows.h` exists?

**Comment by officialmiket** [comment]
I am using ubuntu 21.04. Also i don't have the dir of wine in include if you can help i would be appreciative. I don't really know what to do because after searching windows.h is in wine 1.6-dev that i cannot not download to my knowledge if you have any ideas what i can do it would be helpful.

**Comment by SkrapeProjects** [comment]
**TL;DR**: Try running `sudo apt install wine32-tools wine32-development-tools`, and check if `/usr/include/wine/wine/windows/windows.h` and `/usr/include/wine-development/wine/windows/windows.h` are present.

Little mistake on my part. I am using Fedora, and the `windows.h` header is located in the previously mentioned location, and I assumed it was the same on Ubuntu. It turns out, however, that apparently it's not.
After some testing in a VM with Ubuntu 20.04 installed, I managed to track down the two packages that seem to provide the `windows.h` header, those being `wine32-tools` and `wine32-development-tools`. The new location for the file should be `/usr/include/wine/wine/windows/windows.h` or `/usr/include/wine-development/wine/windows/windows.h`.

> Side note the arrow next to the wine version 5.6 does is mean older or newer?

It technically means you need wine 5.6 or older, but it's based on old information, so that's probably not the problem.

> if you can help i would be appreciative

No problem, I'm here to help you as much as I can ;)

**Comment by officialmiket** [comment]
Am back i have been trying to find for an answer for when i have tried to download wine32-tools it comes up with this 

```
The following packages have unmet dependencies:
 libc6-dev:i386 : Depends: linux-libc-dev:i386 but it is not going to be installed
```
I haven't found any answers from the internet but i tried it in a virtual machine a fresh install it worked perfectly so can i just build it in the virtual machine and take the build to my main machine if not, is there a workaround to the dependencies also i cannot download the linux-libc-dev:i1386 it comes up with some other dependencies that i cannot also download. Like said last time if you can help i would appreciate it.

**Comment by SkrapeProjects** [comment]
I remember running into the same problem. If I remember correctly, I fixed it by installing the packages `gcc-multilib` and `g++-multilib`. 

If that doesn't work you can always try and compile it in a virtual machine and then copy it over to you main, just like you said.
**However** if you're going to do that, it's important to keep in mind that Opentrack won't work if it's not in the path specified by `CMAKE_INSTALL_PREFIX` or in a location that requires root access.

**What I would try to do** is, in the VM, create a fake home directory with the same name of my user on my main machine, with a `.local` directory inside (the command would look something like this `sudo mkdir -p /home/<main_machine_username>/.local && sudo chown -R <vm_username> /home/<main_machine_username>`, I don't know if it has to be run as root but just in case). I would then set `CMAKE_INSTALL_PREFIX` to that location (`/home/<main_machine_username>/.local`), turn on `SDK_WINE`, make and make install.
Then I'd just copy the `.local` folder, located in the fake home directory I created earlier on the VM, to my home directory on my main machine, and in theory that should work. I hope.

**Comment by officialmiket** [comment]
This will maybe be the second to last post because even that you are nice i would like it to getting running but i got i build in the vm, got the .local transferred to my .local but when i try to launch shared libraries (am hoping that it is the correct file) the launching icon just spins and after few seconds it disappears. I don't know if it because i launch it through gui folder, if you know whats the issue or how did you launch the build opentrack and which file did you launch? I would be appreciative if you can help me.

**Comment by SkrapeProjects** [comment]
The file you should be trying to launch is `/home/<your_username>/.local/bin/opentrack`.
You can check if Opentrack works by opening the terminal and typing `sh -c "/home/$(whoami)/.local/bin/opentrack"`.
This will also print out the errors the program might encounter.

**Comment by officialmiket** [comment]
You, i can't put into words how thankful i am. maybe am just to overjoyed because i took me three days but i learnt a lot of new linux commands i have been using linux for one year but i learnt a lot over the three days. I got it working would have been possible without you're help. I wanna say it kept saying permission denied with the command `sh -c "/home/$(whoami)/.local/bin/opentrack"` what i did instead was to cd into the bin dir and from there i did `sh -c opentrack` i got it open. Thank you.

**Comment by SkrapeProjects** [comment]
I'm glad you managed to fix your problem :+1: 
It also took me a while to get it working, that's why I made this guide!

As for the permission denied thing, you might want to try one of these commands:
- `chmod +x /home/<your_username>/.local/bin/opentrack` which flags the file as an executable (it's probably this one)
- `sudo chown <your_username> /home/<your_username>/.local/bin/opentrack` which sets you as the owner of the file

You can also add the `~/.local/bin` directory to your shell path, by running this command `echo "export PATH=\"/home/<your_username>/.local/bin:\$PATH\"" >> ~/.bashrc`, so that you can launch the program by just typing `opentrack` in the terminal (you'll have to relaunch the terminal for the changes to take effect).

**Comment by p-s-t** [comment]
Hello Mark
Thank you for your guide it was something i was looking for for so long.  
I haven't got this far yet in any of my previous attempts.
I too have difficulty finishing. 

when attempting to build with command
make -j$(nproc) 

process stopped around 30 procent 

pointing at 
two errors
i tried starting it again and it did continue always stopping after couple more percent
And eventually it got stuck at 96%  
Unfortunately i cannot scroll too far back to confirm they were just these two, 
but from 5 error messages i can see - it is constantly this on repeat :

/usr/bin/ld: relocatable linking with relocations from format elf32-i386 (/usr/lib/i386-linux-gnu/wine/libwinecrt0.a(exe_entry.o)) to format elf64-x86-64 (opentrack-wrapper-wine.oY6QS3.o) is not supported
winebuild: /usr/bin/ld failed with status 1
winegcc: /usr/lib/wine/winebuild failed
make[2]: *** [proto-wine/CMakeFiles/wine-wrapper.dir/build.make:75: proto-wine/opentrack-wrapper-wine.exe.so] Error 2
make[1]: *** [CMakeFiles/Makefile2:2293: proto-wine/CMakeFiles/wine-wrapper.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
 
i did install this
sudo apt install wine32-tools wine32-development-tools
and 
sudo apt install gcc-multilib g++-multilib

and i looked and do have these two files
/usr/include/wine/wine/windows/windows.h or /usr/include/wine-development/wine/windows/windows.h

But i am stuck again.
Could you please help me ? If you get some spare time.
...also what is SDK_XPLANE ? Will i need it for opentrack to work with xplane ? Or it is some dev kit i wont need ?
Thank you again for writing your guide and helping us.
p-s-t

**Comment by markx86** [comment]
Hi @p-s-t!  
Unfortunately, the guide is broken on Ubuntu based distros (and probably every Debian based distro).  I was already aware of this problem, but due to time and the fact that don't use Ubuntu anymore, I haven't been able to find a solution yet.  

Another thing worth noting are the [limitations](https://github.com/JT8D-17/opentrack#versioninfo) of opentrack's WINE output protocol, mainly the fact that it doesn't work with newer versions of Proton (> 5.9).  

The solution I found is running the Linux version of opentrack (built without WINE support) and setting the output to `UDP over network`, then running the Windows version of opentrack, through WINE, and setting the input to `UDP over network` and the output to `freetrack` (or whatever works on Windows for that specific game). After everything's setup, I press `Start` first in the Windows version of opentrack and then in the Linux version. This adds a very small delay on my System, but it seems to work well enough. I've made a script which also works with Steam games called [opentrack-launcher](https://github.com/VolatileMark/opentrack-launcher), but I haven't used it in a long time, so I don't know if it still works.  

As for the SDK_XPLANE thingy, I honestly don't know... But I would assume it is needed. A quick google search led me to [this page](https://developer.x-plane.com/sdk/plugin-sdk-downloads/) on the developer.x-plane.com website, but I can't seem to get it to build... 

I hope this helps somewhat...
